### DISCUSSION TITLE
Joint Direct/Heuristic/ML Navigator

### DISCUSSION THREAD

**Discussion by asalzburger** [acts-project/acts#741] [Ideas / Feature Requests]
@XiaocongAi @benjaminhuth @paulgessinger @niermann999  - here's something I *think* we could consider.

@benjaminhuth and I discussed the possibility to create a *heuristic* navigator that basically has a trial-and-error machinery behind, i.e. a parser that collects all surfaces (configured to whatever type) from a given tracking geometry via the 

```c++
TrackingGeometry::visitSurface(...)
```

visitor pattern. In every `status()` call it sorts through the surfaces and drops the unreachable ones, but keeps all reachable ones (probably with some bound tolerance). The idea is to get a slow but true ground truth navigation engine.

Now, practically this is nothing else  then the `DirectNavigator` just with the difference that the direct version of it has a predefined sequence of surfaces to be walked through (which is static).

What I am thinking of is a sort of 

```c++
template <typename guider_t>
class GuidedNavigator {

}
```

shell, where one implementation is the `DirectGuider`, one the `HeuristicGuider` and then, when @benjaminhuth has trained his `MLGuider` it could just slit in as a new module. And eventually, when @niermann999  and I have a version of the `DetrayGuider` we could plug it into the same infrastructure.

What do you think?

**Comment by benjaminhuth** [comment]
I think that is a really good idea, because otherwise we would have a lot of redundant/duplicate code implementing the navigator interface for all kinds of navigators.

These `Guider` classes would probably have an interface such as

```c++
class MyGuider
{
   std::vector<const Surface *> operator()(const propagator_state &);
},
```
right?

**Reply by asalzburger** [reply]
Very much so - we could potentially even hide the full current navigator into the same structure at some point.

One question for @XiaocongAi though:
The KalmanFilter is currently the *only* place where we access the ```currentVolume``` and ```currentLayer``` from the Navigator::State.
- do we need that ?

If not, the publicly accessible information of the State becomes very small (`currentSurface` - to which the Action structs listen) and some break condition.

**Reply by asalzburger** [reply]
@benjaminhuth @niermann999  - since this would touch both your projects, I would actually invite you guys to discuss and propose & eventually implement such a setup.

**Reply by benjaminhuth** [reply]
Sure! 
Since I already have a first implementation of my ML navigator, this could potentially serve as a starting point for this.

**Reply by paulgessinger** [reply]
Do we not need `currentVolume` to select the stepper extension?

**Reply by asalzburger** [reply]
Oh, that could be ... we probably should do here is to set the `currentVolumeMaterial` instead of the `currentVolume` ?

**Reply by niermann999** [reply]
I think that sounds great! Especially since we have to decide the memory layout for the surface containers at somepoint

**Reply by XiaocongAi** [reply]
> One question for @XiaocongAi though:
> The KalmanFilter is currently the _only_ place where we access the `currentVolume` and `currentLayer` from the Navigator::State.
> 
> * do we need that ?
> 
We have to reset the `currentVolume` and the `currentLayer` in the (C)KF to reset the navigation status for reversing the navigation direction for the backward filtering in the KF, and for jumping back and forth at different surfaces in the CKF. I'm not sure if this reset of the `currentVolume` and `currentLayer` could be removed when we reset the propagation state.

**Reply by asalzburger** [reply]
Ah, cool, I think we can work around this, with adding a `Navigator::State::reset()` concept as we have for the Stepper.

**Reply by XiaocongAi** [reply]
That's a great idea. It will simplify things a lot.
